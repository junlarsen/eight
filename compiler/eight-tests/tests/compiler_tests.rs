//! End-to-end tests for the compiler with a variety of different inputs and compiler options.

macro_rules! compiler_test_suite {
    ($name:ident, $directory:expr, [$($arg:expr),*]) => {
        #[test]
        fn $name() {
            use std::process::Command;

            insta::with_settings!({
                description => format!("compiler test suite for {}", stringify!($name)),
                snapshot_path => concat!(env!("CARGO_MANIFEST_DIR"), "/tests/snapshots/", stringify!($name)),
            }, {
                insta::glob!(concat!($directory, "/**/*.test"), |path| {
                    let mut cmd = Command::new(insta_cmd::get_cargo_bin("eightc"));
                    $(cmd.arg($arg);)*
                    let relative = path.strip_prefix(env!("CARGO_MANIFEST_DIR"))
                        .expect("failed to build relative path from Cargo.toml");
                    insta_cmd::assert_cmd_snapshot!(cmd.arg(relative));
                });
            })
        }
    }
}

compiler_test_suite!(ui, "ui", []);
compiler_test_suite!(syntax, "syntax", ["--emit-ast"]);
compiler_test_suite!(syntax_lowering, "syntax-lowering", ["--emit-hir"]);
compiler_test_suite!(type_inference, "type-inference", ["--emit-hir"]);
