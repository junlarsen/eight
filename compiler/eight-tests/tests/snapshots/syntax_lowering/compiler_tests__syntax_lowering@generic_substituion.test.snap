---
source: compiler/eight-tests/tests/compiler_tests.rs
description: compiler test suite for syntax_lowering
info:
  program: eightc
  args:
    - "--emit-hir"
    - tests/syntax-lowering/generic_substituion.test
input_file: compiler/eight-tests/tests/syntax-lowering/generic_substituion.test
snapshot_kind: text
---
success: true
exit_code: 0
----- stdout -----
hir_module {
  declares {
    // module types
    
    
    // module struct types
    
    
    // module functions
    fn substitute<T, U, V>(a: i32, b: T, c: U, d: V) -> V;
    
    // module traits
    
    
    // module instances
  
  }
  
  defining {
    // module functions
    fn substitute<$1@0, $1@1, $1@2>(a: i32, b: T, c: U, d: V) -> $1@2 {
      return (d as $1@2);
    }
    // module intrinsic functions
    
  
  }
}

----- stderr -----
